first version documenting how to use the script eessi_container.sh (#100)
EESSI/docs: docs/using_eessi/eessi_container.md

# Using the EESSI container

This page provides all information and guides you through several examples
illustrating to access the EESSI software stack via a small container provided by
EESSI. The container only contains a minimal number tools such as the CernVM-FS
client and some configuration data.

### Prerequisites

> - Singularity 3.7.x or newer _or_ Apptainer x.y.z or newer (any version with

Reviewer: Any Apptainer should work, so change `x.y.z` to `1.0.0`?

Author: Fixed by https://github.com/EESSI/docs/pull/100/commits/0eae16fe58096e7b995d7e7d2ddc3d9208fc2001

  support for `--fusemount` should be sufficient)
- `git` available on your system

### Preparation

> Clone the [`EESSI/software-layer`](https://github.com/EESSI/software-layer.git)
> repository to a directory `eessi_container` by running the command
> ```bash
> git clone https://github.com/EESSI/software-layer.git eessi_container
> ```
> Change directory
> ```bash
> cd eessi_container
> ```

Reviewer: I wouldn't rename `software-layer`, only creates confusion. Also, make it a single box to copy-paste easily?

``````suggestion
Clone the [`EESSI/software-layer`](https://github.com/EESSI/software-layer.git)
repository, and change into the resulting `software-layer` directory:
```bash
git clone https://github.com/EESSI/software-layer.git
cd software-layer
```
``````

Author: Fixed by https://github.com/EESSI/docs/pull/100/commits/0eae16fe58096e7b995d7e7d2ddc3d9208fc2001

## Basic scenarios

> ### Get access to EESSI

Reviewer: ```suggestion
### Getting access to EESSI

Author: Fixed by https://github.com/EESSI/docs/pull/100/commits/0eae16fe58096e7b995d7e7d2ddc3d9208fc2001

Simply run the command (from the directory you have changed into above)
```bash
./eessi_container.sh
```
_(Note, the first run may take a bit longer because the container image is
downloaded and converted.)_

You should see output like
```
Using /tmp/eessi.RC6kYEjmSj as tmp storage (add '--resume /tmp/eessi.RC6kYEjmSj' to resume where this session ended).
Launching container with command (next line):
singularity shell --fusemount container:cvmfs2 pilot.eessi-hpc.org /cvmfs/pilot.eessi-hpc.org docker://ghcr.io/eessi/build-node:debian11
CernVM-FS: pre-mounted on file descriptor 3
Singularity> CernVM-FS: loading Fuse module... done
fuse: failed to clone device fd: Inappropriate ioctl for device
fuse: trying to continue without -o clone_fd.

Singularity>
```

> _(Note, you may have to press enter to clearly see the prompt as some messages

Reviewer: For notes, you can use:

```
!!! note
    Example
```

Author: Fixed by https://github.com/EESSI/docs/pull/100/commits/0eae16fe58096e7b995d7e7d2ddc3d9208fc2001

Also changed most _Note ..._ accordingly.

`Singularity> ` was shown.)_
To access the EESSI pilot stack just run the command
```
source /cvmfs/pilot.eessi-hpc.org/latest/init/bash
```
This may take a while as data is downloaded from Stratum 1 server which is part of the
CernVM-FS infrastructure to distribute files. You should see the following (or
similar output depending on the CPU architecture of the machine you are accessing
EESSI).
```
Found EESSI pilot repo @ /cvmfs/pilot.eessi-hpc.org/versions/2021.12!
archspec says x86_64/intel/skylake_avx512
Using x86_64/intel/skylake_avx512 as software subdirectory.
Using /cvmfs/pilot.eessi-hpc.org/versions/2021.12/software/linux/x86_64/intel/skylake_avx512/modules/all as the directory to be added to MODULEPATH.
Found Lmod configuration file at /cvmfs/pilot.eessi-hpc.org/versions/2021.12/software/linux/x86_64/intel/skylake_avx512/.lmod/lmodrc.lua
Initializing Lmod...
Prepending /cvmfs/pilot.eessi-hpc.org/versions/2021.12/software/linux/x86_64/intel/skylake_avx512/modules/all to $MODULEPATH...
Environment set up to use EESSI pilot software stack, have fun!
[EESSI pilot 2021.12] $ 
```
The last line is the prompt. Run `module avail` to see which modules and

> extensions are available. About 280 modules and 1800 extensions are installed.

Reviewer: I wouldn't mention numbers here, nightmare to keep up-to-date...

Author: Removed by https://github.com/EESSI/docs/pull/100/commits/0eae16fe58096e7b995d7e7d2ddc3d9208fc2001

   RStudio-Server/1.3.1093-foss-2020a-Java-11-R-4.0.0
```
Load modules with `module load package/version`, e.g.,
`module load R/4.1.0-foss-2021a`, and try out the software. See below for a short
session
```
[EESSI pilot 2021.12] $ module load R/4.1.0-foss-2021a
[EESSI pilot 2021.12] $ which R
/cvmfs/pilot.eessi-hpc.org/versions/2021.12/software/linux/x86_64/intel/skylake_avx512/software/R/4.1.0-foss-2021a/bin/R
[EESSI pilot 2021.12] $ R --version
R version 4.1.0 (2021-05-18) -- "Camp Pontanezen"
Copyright (C) 2021 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under the terms of the
GNU General Public License versions 2 or 3.
For more information about these matters see
https://www.gnu.org/licenses/.
```
Leave the container with `exit`.
### Reusing the previous session
You may have noted the following line in the output of `eessi_container.sh`
```
Using /tmp/eessi.RC6kYEjmSj as tmp storage (add '--resume /tmp/eessi.RC6kYEjmSj' to resume where this session ended).
```
Just try the following command to "resume" from the last session. _Note, the

> parameter after `--resume` is very likely different for your last session. Just

Reviewer: `is very likely` -> `*will* be`

Author: Changed by https://github.com/EESSI/docs/pull/100/commits/0eae16fe58096e7b995d7e7d2ddc3d9208fc2001

scroll back and take note of it._
```

> ./eessi_container.sh --resume /tmp/eessi.RC6kYEjmSj

Reviewer: Maybe we should use a random bit here that's clearly an example, something like `ABC123defg`?

Author: Used `/tmp/eessi.EXAMPLE` instead.

```
This should run much faster because the container image has been cached in the
temporary directory (`/tmp/eessi.RC6kYEjmSj` in the example above). You should
get to the prompt (`Singularity > ` or `Apptainer > `). Now you just have
to source again the EESSI init script by running
```
source /cvmfs/pilot.eessi-hpc.org/latest/init/bash
```

> _(Note, the `/tmp/eessi...` directory contains a `home` directory which includes

Reviewer: ```
!!! note
    The `/tmp/...` ...

Author: Fixed by https://github.com/EESSI/docs/pull/100/commits/0eae16fe58096e7b995d7e7d2ddc3d9208fc2001

                          repository to use [default: EESSI-pilot via
                          container configuration]
  -u | --resume DIR/TGZ - resume a previous run from a directory or tarball,
                          where DIR points to a previously used tmp directory
                          (check for output 'Using DIR as tmp ...' of a previous
                          run) and TGZ is the path to a tarball which is
                          unpacked the tmp dir stored on the local storage space
                          (see option --storage above) [default: not set]
  -s | --save DIR/TGZ   - save contents of tmp directory to a tarball in
                          directory DIR or provided with the fixed full path TGZ
                          when a directory is provided, the format of the
                          tarball's name will be {REPO_ID}-{TIMESTAMP}.tgz
                          [default: not set]
  -v | --verbose        - display more information [default: false]

 If value for --mode is 'run', the SCRIPT provided is executed.
```
## Running commands or scripts with the container
### Run a command
### Run a script
### Run EESSI demo scripts

> ## Build a package for `/cvmfs/pilot.eessi-hpc.org`

Reviewer: Maybe something like `## Getting (local) write access to the EESSI repository`, and then clarify what this means?

Reviewer: We should definitely also have a section on save/resume

Reviewer: Fine by me to cover `--save` in a later PR
